lms/djangoapps/courseware/management/commands/fix_student_module_newlines.py

```diff
@@ -0,0 +1,330 @@
+        """
+        Fix a StudentModule with a trailing newline in the course_id.
+
+        Returns a count of (row modified, )
+
+        At the end of this method call, the record should no longer have a
+        trailing newline for the course_id. There are three possible outcomes:
+
+        1. There was never a conflicting entry:
+            -> We just update this row.
+        2. Conflict and the other row (with correct course_id) wins:
+            -> We delete this row.
+        2. Conflict and this row wins:
+            -> We copy the data to the conflicting row (the one that has a
+               correct course_id), and delete this row.
+
+        Even though all the StudentModule entries coming in here have trailing
+        newlines in the course_id, the deserialization logic will obscure that
+        (it gets parsed out when read from the database). We will also
+        automatically strip the newlines when writing back to the database, so
+        we have to be very careful about violating unique constraints by doing
+        unintended updates. That's why we only do an update to
+        newline_course_row if no correct_course_row exists.
+        """
+        # pylint: disable=too-many-statements
+        # We got the StudentModule from the read replica, so we have to fetch it
+        # again from our writeable database before making changes
+        try:
+            newline_course_row = StudentModule.objects.get(id=read_only_newline_course_row.id)
+        except StudentModule.DoesNotExist:
+            # We're not going to be able to make any corrective writes, so just fail fast
+            log.exception("Could not find editable CSM row %s", read_only_newline_course_row.id)
+            return FixResult(record_trimmed=0, data_copied=0, record_deleted=0, capa_merge=0, error=1)
+
+        # Find the StudentModule entry with the correct course_id.
+        try:
+            correct_course_row = StudentModule.objects.get(
+                student_id=newline_course_row.student_id,
+                course_id=newline_course_row.course_id,
+                module_state_key=newline_course_row.module_state_key,
+            )
+            assert correct_course_row.pk != newline_course_row.pk
+        except StudentModule.DoesNotExist:
+            correct_course_row = None
+
+        # If only an entry with the newline course exists, just change the course_id and save.
+        if correct_course_row is None:
+            log.info(
+                "No conflict: Removing trailing newline from course_id in CSM row %s - (%s) %s",
+                newline_course_row.id,
+                newline_course_row.module_type,
+                newline_course_row.module_state_key,
+            )
+            if dry_run:
+                return FixResult(record_trimmed=1, data_copied=0, record_deleted=0, capa_merge=0, error=0)
+
+            try:
+                with transaction.atomic():
+                    newline_course_row.save()
```

Comment:
I think we get that with the transaction mode anyway. And if the student does update the row in the meantime (it has to be the correct course key row, since that's all they can access), and it causes it to fail, then that's ok. We'd want to keep the old one anyway. And we can look it up in the log and manually kill what failed if necessary.